pulsar-transaction/buffer/src/main/java/org/apache/pulsar/transaction/buffer/impl/PersistentTransactionBuffer.java

```diff
@@ -0,0 +1,308 @@
+                    return null;
+                }).exceptionally(e -> {
+                    commitFuture.completeExceptionally(e);
+                    return null;
+                });
+                return null;
+            }).exceptionally(e -> {
+                commitFuture.completeExceptionally(e);
+                return null;
+            });
+            return null;
+        }).exceptionally(e -> {
+            commitFuture.completeExceptionally(e);
+            return null;
+        });
+
+        return commitFuture;
+    }
+
+    @Override
+    public CompletableFuture<Void> abortTxn(TxnID txnID) {
+        CompletableFuture<Void> abortFuture = new CompletableFuture<>();
+
+        Marker abortMarker = Marker.builder().txnID(txnID).status(TxnStatus.ABORTED).build();
+        ByteBuf marker = Unpooled.wrappedBuffer(abortMarker.serialize());
+
+        txnCursor.getTxnMeta(txnID, false).thenCompose(meta -> {
+            publishMessage(marker, meta.lastSequenceId() + 1).thenCompose(position -> {
+                txnCursor.abortTxn(txnID).thenCompose(v -> {
+                    abortFuture.complete(null);
+                    return null;
+                }).exceptionally(e -> {
+                    abortFuture.completeExceptionally(e);
+                    return null;
+                });
+                return null;
+            }).exceptionally(e -> {
+                abortFuture.completeExceptionally(e);
+                return null;
+            });
+            return null;
+        }).exceptionally(e -> {
+            abortFuture.completeExceptionally(e);
+            return null;
+        });
+
+        return abortFuture;
+    }
+
+    private CompletableFuture<Position> publishMessage(ByteBuf msg, long sequenceId) {
+        CompletableFuture<Position> publishFuture = new CompletableFuture<>();
+
+        CountDownLatch latch = new CountDownLatch(1);
+
+        brokerService.executor().execute(() -> {
+            publishMessage(msg, new TxnCtx(sequenceId) {
+                @Override
+                public void completed(Exception e, long ledgerId, long entryId) {
+                    if (e != null) {
+                        publishFuture.completeExceptionally(e);
+                        return;
+                    }
+
+                    publishFuture.complete(PositionImpl.get(ledgerId, entryId));
+                    latch.countDown();
+                }
+            });
+        });
+
+        try {
+            latch.await();
+        } catch (InterruptedException e) {
+            publishFuture.completeExceptionally(e);
+        }
+
+        return publishFuture;
+    }
+
+    @Override
+    public CompletableFuture<Void> purgeTxns(List<Long> dataLedgers) {
+        CompletableFuture<Void> purgeFuture = new CompletableFuture<>();
+
+        if (log.isDebugEnabled()) {
+            log.debug("Begin to purge the ledgers {}", dataLedgers);
+        }
+
+
+        dataLedgers.forEach(dataLedger -> {
```

Comment:
purgeFuture need to wait all dataLedgers are deleted. but the implementation here doesn't guarantee it.